src/wallet/wallet2.cpp

```diff
@@ -12935,6 +13001,188 @@ bool wallet2::import_key_images(signed_tx_set & signed_tx, size_t offset, bool o
+        dbd.detached_confirmed_txs_dests.find(bgs_tx.first) != dbd.detached_confirmed_txs_dests.end())
+    {
+      m_confirmed_txs[bgs_tx.first].m_dests = std::move(dbd.detached_confirmed_txs_dests[bgs_tx.first]);
+    }
+  }
+
+  reattach_blockchain(m_blockchain, dbd);
+
+  THROW_WALLET_EXCEPTION_IF(!m_background_sync_cache.empty(), error::wallet_internal_error,
+      "Expected to process all background syncd txs");
+
+  LOG_PRINT_L2("Finished processing background sync cache");
+}
+//----------------------------------------------------------------------------------------------------
+/*
+  When background sync is enabled, the wallet scans using just the view key. This
+  enables a client to sync in the background, without keeping the spend key in
+  decrypted state in memory. When a user returns to the wallet and decrypts the
+  spend key, the wallet processes the background sync'd txs, then the wallet
+  picks up scanning normally right where the background sync left off.
+*/
+void wallet2::enable_background_sync()
+{
+  LOG_PRINT_L0("Enabling background sync");
+
+  THROW_WALLET_EXCEPTION_IF(m_multisig || m_watch_only || key_on_device(), error::wallet_internal_error, m_multisig
+      ? "Background sync not implemented for multisig wallets" : m_watch_only
+      ? "Background sync not implemented for view only wallets"
+      : "Background sync not implemented for HW wallets");
+
+  if (m_background_sync_enabled)
+    return;
+  m_background_sync_enabled = true;
+
+  // Wipe the spend key from memory. Will bring it back when background sync
+  // mode is disabled.
+  m_account.forget_spend_key();
+
+  LOG_PRINT_L0("Background sync enabled");
+}
+//----------------------------------------------------------------------------------------------------
+void wallet2::disable_background_sync(const epee::wipeable_string &password, const crypto::secret_key &spend_secret_key)
+{
+  LOG_PRINT_L0("Disabling background sync");
+
+  THROW_WALLET_EXCEPTION_IF(m_multisig || m_watch_only || key_on_device(), error::wallet_internal_error, m_multisig
+      ? "Background sync not implemented for multisig wallets" : m_watch_only
+      ? "Background sync not implemented for view only wallets"
+      : "Background sync not implemented for HW wallets");
+
+  // Verify provided password and recover spend secret key from wallet file
+  crypto::secret_key recovered_spend_key = crypto::null_skey;
+  THROW_WALLET_EXCEPTION_IF(!m_wallet_file.empty() && !verify_password(password, recovered_spend_key),
+      error::invalid_password);
+
+  if (spend_secret_key != crypto::null_skey)
+  {
+    THROW_WALLET_EXCEPTION_IF(!m_wallet_file.empty() && spend_secret_key != recovered_spend_key,
+        error::invalid_spend_key);
+    LOG_PRINT_L0("Setting spend secret key with the provided key");
+    recovered_spend_key = spend_secret_key;
+  }
+
+  // Verify private spend key derives to wallet's public spend key
+  crypto::public_key spend_public_key;
```

Comment:
Well, at least in theory we still support non-deterministic wallets where we have completely independent spend and view keys ... maybe just forbid background sync for them and all is well?

Reply:
This is checking the private spend key derives to the public spend key (not that the view key is derived from spend key). Do you know of any circumstances where this isn't the case except for with multisig wallets (which can't background sync)?